### PR TITLE
Add support for custom docs layout with document

### PR DESCRIPTION
**Pull request by sunnylqm**
This pr is a successor of https://github.com/facebook/Docusaurus/pull/128, with an additional `default` config. Also added documents requested in https://github.com/facebook/Docusaurus/issues/737.

### CODE REVIEW

**Comment by sunnylqm** [comment]
@yangshun Fixed. But why the cicleci is on hold?

File: docs/api-site-config.md [hunk 1]
```diff
@@ -143,6 +143,33 @@ h1 {
 * `defaultLang` defines a default language. It will be used if one is not specified at the top of the code block. You can find the [list of supported languages here](https://github.com/isagalaev/highlight.js/tree/master/src/languages).
 * `themeUrl` is the custom URL of CSS theme file that you want to use with Highlight.js. If this is provided, the `theme` and `version` fields will be ignored.
 
+`layouts` - You can specify some custom components here to override the [offcial layout template](https://github.com/facebook/Docusaurus/blob/master/lib/core/Doc.js) to render some or all of the doc pages. 
+
+```js
+layouts: {
+  // default will be applied to all doc pages
+  default: function({ React, MarkdownBlock }) {
```
Comment (yangshun) [inline docs/api-site-config.md]: I don't think is true. From reading the code, I don't think we have a way now to specify a default layout for all pages. This is simply a layout called `default` without any special functionality.

Comment (sunnylqm) [reply docs/api-site-config.md]: https://github.com/facebook/Docusaurus/pull/763/files#diff-fe37ce87a7b7b83106303bf9233b49fbR238

Comment (yangshun) [reply docs/api-site-config.md]: My bad, I missed out on that chunk. I was looking at existing code as I had the impression this PR was only documenting the layouts behavior.

Comment (endiliey) [reply docs/api-site-config.md]: Actually i believe PR #128 is incomplete. It is only available at dev server, not on build. Might need to modify

https://github.com/facebook/Docusaurus/blob/c2ff41365039c2d86f9d5d6da0370cae31ab0a32/lib/server/generate.js#L195

Comment (sunnylqm) [reply docs/api-site-config.md]: @endiliey Added.

Comment (yangshun) [reply docs/api-site-config.md]: Good catch @endiliey! We really need to reduce the double work in server.js and generate.js

File: lib/core/DocsLayout.js [hunk 2]
```diff
@@ -30,15 +30,29 @@ class DocsLayout extends React.Component {
...
-      DocComponent = this.props.Doc;
-    }
+
     const title = i18n
       ? translation[this.props.metadata.language]['localized-strings'][
           this.props.metadata.localized_id
         ] || this.props.metadata.title
       : this.props.metadata.title;
+
+    const config = this.props.siteConfig;
```
Comment (endiliey) [inline lib/core/DocsLayout.js]: You are passing siteConfig as config props

```js
<DocsLayout metadata={metadata} language={language} config={siteConfig}>
```

It should be
```js
const config = this.props.config;
```

File: lib/core/DocsLayout.js [hunk 3]
```diff
@@ -34,15 +34,26 @@ class DocsLayout extends React.Component {
...
           this.props.metadata.localized_id
         ] || this.props.metadata.title
       : this.props.metadata.title;
+    const customLayouts = this.props.config.layouts;
+    let DocComponent = Doc;
+    if (customLayouts) {
+      if (customLayouts.default) {
+        DocComponent = customLayouts.default({
+          React,
+          MarkdownBlock,
```
Comment (endiliey) [inline lib/core/DocsLayout.js]: I think you forgot to import MarkdownBlock

You could do something like 
```js
MarkdownBlock: require('./MarkdownBlock.js')
```

or import it beforehand
```js
const MarkdownBlock = require('./MarkdownBlock.js');
```

**Comment by endiliey** [comment]
Closing this in favor of v2 implementation

**Pull request closed without merging**